Add Cancel/Unknown/Clear Hardware Keycodes.

jameshilliard opened the pull request:
#### acb9bae813d5cf135288230344d3b4b22818e6c8
<pre>
Add Cancel/Unknown/Clear Hardware Keycodes.
<a href="https://bugs.webkit.org/show_bug.cgi?id=251778">https://bugs.webkit.org/show_bug.cgi?id=251778</a>

Reviewed by Adrian Perez de Castro.

This adds handling for the following keycodes:
KEY_CANCEL
KEY_UNKNOWN
KEY_CLEAR

* Source/WebCore/platform/libwpe/PlatformKeyboardEventLibWPE.cpp:
(WebCore::PlatformKeyboardEvent::keyCodeForHardwareKeyCode):

Canonical link: <a href="https://commits.webkit.org/261008@main">https://commits.webkit.org/261008@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/402c14d46284b5b0f96633b05c88ff050614f4e1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106299 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15353 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39138 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115492 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/115179 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110209 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16795 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6569 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98515 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112063 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/12779 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95761 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/40339 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94654 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27404 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/82033 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8593 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28756 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9103 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5328 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14716 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48302 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7627 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10639 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebCore/platform/libwpe/PlatformKeyboardEventLibWPE.cpp:
```diff
@@ -758,6 +758,12 @@ String PlatformKeyboardEvent::keyCodeForHardwareKeyCode(unsigned keyCode)
+    case 0x00F8:
+        return "Unknown"_s;
+    case 0x016B:
+        return "Clear"_s;
```
Where do these codes come from? I have not been able to find them in the MDN pages linked from the comments in this file.

jameshilliard replied:
You just take the corresponding [input event key code](https://github.com/freedesktop/libevdev/blob/libevdev-1.13.0/include/linux/linux/input-event-codes.h) and [add 8 to it](https://github.com/Igalia/cog/blob/0.16.1/platform/wayland/cog-platform-wl.c#L1212-L1214) to get an equivalent MDN key code. These new key codes are not currently listed in the MDN pages so I just named them based off of the corresponding key event code define names.

aperezdc replied:
Ah, I saw now your comment [in bug #251778](https://bugs.webkit.org/show_bug.cgi?id=251778)... Not sure how much we can we expect there will be other devices which use the same codes. Also I wonder why your device would not produce `WPE_KEY_Cancel` instead of `0x00E7` 🤔

jameshilliard replied:
> Not sure how much we can we expect there will be other devices which use the same codes. 

They are [kernel defined key codes](https://github.com/torvalds/linux/blob/v6.2-rc7/include/uapi/linux/input-event-codes.h) I guess so I presume there are other devices using them.

> Also I wonder why your device would not produce `WPE_KEY_Cancel` instead of `0x00E7` 🤔

It produces both, [`WPE_KEY_Cancel`](https://github.com/WebPlatformForEmbedded/libwpe/blob/f4cbc1ff330d31f8761585bd392e8d093257c0bc/include/wpe/keysyms.h#L202) is the keysym not the hardware key code:
```
handle_key_event key: 0x000000e7 keysym: 0x0000ff69
```

jameshilliard committed:
Add Cancel/Unknown/Clear Hardware Keycodes.
https://bugs.webkit.org/show_bug.cgi?id=251778

Reviewed by Adrian Perez de Castro.

This adds handling for the following keycodes:
KEY_CANCEL
KEY_UNKNOWN
KEY_CLEAR

* Source/WebCore/platform/libwpe/PlatformKeyboardEventLibWPE.cpp:
(WebCore::PlatformKeyboardEvent::keyCodeForHardwareKeyCode):

Canonical link: https://commits.webkit.org/261008@main
